### PR TITLE
Make invocation number accessible

### PR DESCRIPTION
**Pull request by naomiGrew**

### CODE REVIEW

**Comment by naomiGrew** [comment]
fixes #216

File: docs/config.md [hunk 1]
```diff
@@ -458,7 +459,7 @@ Example:
 ```yaml
 benchmark_suites:
   ExampleSuite:
-    command: Harness %(benchmark)s --problem-size=%(input)s --iterations=%(iterations)s
+    command: Harness %(benchmark)s --problem-size=%(input)s --iterations=%(iterations)s %(invocations)
```
Comment (smarr) [inline docs/config.md:462]:
```suggestion
    command: Harness %(benchmark)s --problem-size=%(input)s --iterations=%(iterations)s %(invocation)s
```

File: rebench/executor.py [hunk 2]
```diff
@@ -492,6 +492,9 @@ def _generate_data_point(self, cmdline, gauge_adapter, run_id,
                              termination_check):
         # execute the external program here
         output = ""
+        if 'invocation ' in cmdline:
```
Comment (smarr) [inline rebench/executor.py:495]: Remove the `if`

File: rebench/executor.py [hunk 3]
```diff
@@ -492,6 +492,9 @@ def _generate_data_point(self, cmdline, gauge_adapter, run_id,
...
+            cmdline = cmdline % {'invocation':run_id.completed_invocations +1}
+            run_id.store_cmdline(cmdline)
```
Comment (smarr) [inline rebench/executor.py:497]:
```suggestion
```

File: rebench/model/run_id.py [hunk 4]
```diff
@@ -47,6 +47,7 @@ def __init__(self, benchmark, cores, input_size, var_value, machine):
         self.executable = None
         self.executable_missing = False
         self.is_failed = True
+        self.updated_cmdline = []
```
Comment (smarr) [inline rebench/model/run_id.py:50]:
```suggestion
```

File: rebench/model/run_id.py [hunk 5]
```diff
@@ -272,6 +274,12 @@ def cmdline(self):
             return self._cmdline
         return self._construct_cmdline()
 
+    def store_cmdline(self, cmd):
+        self.updated_cmdline.append(cmd)
+
+    def get_cmdline(self):
+        return self.updated_cmdline
+
```
Comment (smarr) [inline rebench/model/run_id.py:282]:
```suggestion
```

File: rebench/tests/features/issue_216.data [hunk 6]
```diff
@@ -0,0 +1 @@
+
```
Comment (smarr) [inline rebench/tests/features/issue_216.data:1]: please delete file

**Review by smarr — CHANGES_REQUESTED** [review]

**Comment by naomiGrew** [comment]
config.md needs an example of how to use %(invocations) around line 220, i have added a description but i couldn't think of a good example

**Comment by naomiGrew** [comment]
The test currently only results on one data point being produced?
/Users/naomigrew/Documents/ReBench/rebench/tests/features/issue_216_test.py:45: in _records_data_points
    self.assertEqual(num_data_points, run.get_number_of_data_points())
E   AssertionError: 10 != 1

**Comment by naomiGrew** [comment]
for loop in issue_216_vm.py is needed otherwise it causes the following error 
self = <rebench.model.data_point.DataPoint object at 0x103588a90>

    def get_total_unit(self):
>       return self._total.unit
E       AttributeError: 'NoneType' object has no attribute 'unit'

/Users/naomigrew/Documents/ReBench/rebench/model/data_point.py:55: AttributeError

**Comment by smarr** [comment]
Thanks for preparing this.

I added an example to the documentation, simplified your test quite a bit so that it really just prints out the current invocation number, and configure it to invoke the run 4 times, which checks what we want to check, and makes the test pass.

Though, I noticed that we can't really just expand the variable in the executor, because this is after we appended all kind of other, possibly arbitrary data to the command line.
For instance, the TimeAdapter allows us to measure the maximum amount of memory being used, which inserts a completely unrelated format string with a `%` into the command line. So, I actually moved this to the `RunId` and make all these adapters call the new `run_id.cmdline_for_next_invocation()` method, before the adapters add their stuff. This way, we can be sure there's nothing else incompatible in the command line.

**Pull request merged**